### ISSUE TITLE
`nan` loss with Tensorflow Datasets

### ISSUE THREAD

**Issue by shivance** [keras-team/keras-nlp#880]
Related Issue : #726 , [keras-io/#1253](https://github.com/keras-team/keras-io/pull/1253)
 
Notebook Link : https://colab.research.google.com/drive/1KCNxATaeNcmxk-XJcgYBAFb4gngk5_rf?usp=sharing

Working on Comments by @mattdangerw , I tried to remove the usage of pandas conversion and directly creating tf.data.Dataset from the data that was downloaded through `tensorflow-datasets` . I tried training BERT and RoBERTa classifier on the data, but it is displaying nan loss.

Previously, when I was using pandas, and then preprocessing the columns, and was fitting models by passing x and y explicitly, and with the bert tiny classifier around ~68% accuracy was achieved. Now it's stuck at 32% throughout

As far as I can figure out, the only thing changed here is data pipeline ( preprocessing data and fitting the tf dataset). Where am I doing wrong?

![image](https://user-images.githubusercontent.com/51750587/226174039-061a11b6-cc30-4c81-aea0-a6284ebd6dfc.png)

Notebook outputs display the nan losses. 

PS: I'm not sure if it deserves a bug label

[Edit] : Linking a colab, to avoid confusions in different version of notebook version on kaggle.

**Comment by shivance** [comment]
Tagging @chenmoneygithub, @fchollet and @abheesht17 for quick help [ wasted 2 hrs, couldn't figure out 🤕 .

**Comment by mattdangerw** [comment]
Good question! I think this boils down to proper usage of the `tf.data` API, which can be a bit tricky and take some getting used to. Here's a colab showing a fixed up input pipeline:

https://colab.research.google.com/gist/mattdangerw/8003a5ba4438b1cccc56a79cbd5712ec/semantic-similarity.ipynb

More detailed notes:

You don't want to do `hypothesis = [sample['hypothesis'].numpy().decode('utf-8') for sample in dataset]`. This will read the whole dataset into memory, and go against what `tf.data` is trying to do. Basically whenever you run a `map` or `batch` or `filter` on your dataset, you aren't actually applying that operation to the dataset right away, you are setting up a pipeline that will run on your raw data when you call `fit()` or `predict()`. The goal is to set up a series of dataset manipulating operations, all within the `tf.data` API, from data loading to data consuming. This guide is super useful -> https://www.tensorflow.org/guide/data_performance

You need to make sure to filter out all the `-1` labels in your dataset. I think this might have been the cause of the `nan`s. The loss function won't accept a -1 label. Another option would be to predict this missing label as another class, say, by adding one to all your labels, and predicting 4 classes. But I believe the norm with SNLI is to filter these examples with missing labels out.

**Comment by mattdangerw** [comment]
I believe things are all working now on the keras.io tutorial, so will go ahead and close this!

**Comment by shivance** [comment]
Thanks @mattdangerw for the help here, forgot to mention it, 
It's my bad, while I was porting preprocessing pipeline, I totally missed filtering out -1's , while it was there in earlier version of PR. 

Sorry !
